File: frontend/developer-workstation/src/__tests__/fieldNameValidation.property.test.ts
import { describe, test, expect } from 'vitest'
import * as fc from 'fast-check'

/**
 * Property-Based Tests for Field Name Validation
 * Feature: process-task-form-separation
 *
 * Property 3: Form field names must reference Data_Table columns
 * **Validates: Requirements 2.5, 3.3, 3.4**
 */

/**
 * Validates that a set of field names are all present in the Data_Table columns.
 * Returns the list of invalid field names (not found in columns).
 */
function validateFieldNames(fieldNames: string[], dataTableColumns: string[]): string[] {
  const columnSet = new Set(dataTableColumns)
  return fieldNames.filter(name => !columnSet.has(name))
}

describe('Property 3: Form field names must reference Data_Table columns', () => {
  test('property: field names that exist in Data_Table columns are accepted', () => {
    fc.assert(
      fc.property(
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 1, maxLength: 20 }),
        (columns: string[]) => {
          // Pick a subset of columns as field names — all should be valid
          const fieldNames = columns.slice(0, Math.max(1, Math.floor(columns.length / 2)))
          const invalid = validateFieldNames(fieldNames, columns)
          expect(invalid).toEqual([])
        },
      ),
      { numRuns: 100 },
    )
  })

  test('property: field names not in Data_Table columns are rejected', () => {
    fc.assert(
      fc.property(
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 1, maxLength: 10 }),
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 1, maxLength: 10 }),
        (columns: string[], extraNames: string[]) => {
          // Filter extraNames to only those NOT in columns
          const columnSet = new Set(columns)
          const invalidFieldNames = extraNames.filter(n => !columnSet.has(n))
          if (invalidFieldNames.length === 0) return // skip if all happen to be in columns

          const result = validateFieldNames(invalidFieldNames, columns)
          expect(result.length).toBe(invalidFieldNames.length)
          for (const name of invalidFieldNames) {
            expect(result).toContain(name)
          }
        },
      ),
      { numRuns: 100 },
    )
  })

  test('property: mixed valid and invalid field names — only invalid ones rejected', () => {
    fc.assert(
      fc.property(
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 2, maxLength: 15 }),
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 1, maxLength: 10 }),
        (columns: string[], extraNames: string[]) => {
          const columnSet = new Set(columns)
          const validSubset = columns.slice(0, Math.max(1, Math.floor(columns.length / 2)))
          const invalidSubset = extraNames.filter(n => !columnSet.has(n))

          const allFieldNames = [...validSubset, ...invalidSubset]
          const result = validateFieldNames(allFieldNames, columns)

          // Only the invalid ones should be returned
          expect(result.length).toBe(invalidSubset.length)
          for (const name of validSubset) {
            expect(result).not.toContain(name)
          }
          for (const name of invalidSubset) {
            expect(result).toContain(name)
          }
        },
      ),
      { numRuns: 100 },
    )
  })

  test('property: empty field names list always passes validation', () => {
    fc.assert(
      fc.property(
        fc.uniqueArray(fc.string({ minLength: 1, maxLength: 50 }), { minLength: 0, maxLength: 10 }),
        (columns: string[]) => {
          const result = validateFieldNames([], columns)
          expect(result).toEqual([])
        },
      ),
      { numRuns: 100 },
    )
  })
})

File: backend/developer-workstation/src/test/java/com/developer/property/FieldNameValidationPropertyTest.java
package com.developer.property;

import net.jqwik.api.*;

import java.util.*;
import java.util.stream.Collectors;

import static org.assertj.core.api.Assertions.*;

/**
 * 字段名校验属性测试
 * Feature: process-task-form-separation, Property 3: Form field names must reference Data_Table columns
 *
 * Validates: Requirements 2.5, 3.3, 3.4, 4.5
 */
public class FieldNameValidationPropertyTest {

    /**
     * Property 3: Field names that exist in Data_Table columns should pass validation.
     *
     * For any form field name that exists as a column name in the FunctionUnit's Data_Table,
     * validation should accept it.
     *
     * Validates: Requirements 2.5, 3.3, 3.4, 4.5
     */
    @Property(tries = 100)
    @Label("Feature: process-task-form-separation, Property 3: Valid field names are accepted")
    void validFieldNamesShouldBeAccepted(
            @ForAll("validFieldNameScenarios") FieldNameScenario scenario) {

        // All field names exist in the column set
        for (String fieldName : scenario.fieldNames) {
            assertThat(scenario.dataTableColumns).contains(fieldName);
        }

        List<String> invalidFields = validateFieldNames(scenario.fieldNames, scenario.dataTableColumns);
        assertThat(invalidFields).isEmpty();
    }

    /**
     * Property 3: Field names not in Data_Table columns should be rejected.
     *
     * For any field name that does not exist in the FunctionUnit's Data_Table columns,
     * validation should reject it.
     *
     * Validates: Requirements 2.5, 3.3, 3.4, 4.5
     */
    @Property(tries = 100)
    @Label("Feature: process-task-form-separation, Property 3: Invalid field names are rejected")
    void invalidFieldNamesShouldBeRejected(
            @ForAll("invalidFieldNameScenarios") FieldNameScenario scenario) {

        // At least one field name does not exist in the column set
        boolean hasInvalid = scenario.fieldNames.stream()
                .anyMatch(name -> !scenario.dataTableColumns.contains(name));
        assertThat(hasInvalid).isTrue();

        List<String> invalidFields = validateFieldNames(scenario.fieldNames, scenario.dataTableColumns);
        assertThat(invalidFields).isNotEmpty();
    }

    /**
     * Property 3: The number of rejected fields equals the count of names not in columns.
     *
     * Validates: Requirements 2.5, 3.3, 3.4, 4.5
     */
    @Property(tries = 100)
    @Label("Feature: process-task-form-separation, Property 3: Rejected count matches invalid field count")
    void rejectedCountMatchesInvalidFieldCount(
            @ForAll("mixedFieldNameScenarios") FieldNameScenario scenario) {

        long expectedInvalidCount = scenario.fieldNames.stream()
                .filter(name -> !scenario.dataTableColumns.contains(name))
                .count();

        List<String> invalidFields = validateFieldNames(scenario.fieldNames, scenario.dataTableColumns);
        assertThat(invalidFields).hasSize((int) expectedInvalidCount);
    }

    // ========== Validation Logic ==========

    /**
     * Business rule: All form field names must reference Data_Table columns.
     * Returns the list of field names that are NOT in the Data_Table columns.
     */
    private List<String> validateFieldNames(List<String> fieldNames, Set<String> dataTableColumns) {
        return fieldNames.stream()
                .filter(name -> !dataTableColumns.contains(name))
                .collect(Collectors.toList());
    }

    // ========== Data Classes ==========

    static class FieldNameScenario {
        final List<String> fieldNames;
        final Set<String> dataTableColumns;

        FieldNameScenario(List<String> fieldNames, Set<String> dataTableColumns) {
            this.fieldNames = fieldNames;
            this.dataTableColumns = dataTableColumns;
        }
    }

    // ========== Arbitraries ==========

    @Provide
    Arbitrary<FieldNameScenario> validFieldNameScenarios() {
        return dataTableColumnSets().flatMap(columns -> {
            if (columns.isEmpty()) {
                return Arbitraries.just(new FieldNameScenario(List.of(), columns));
            }
            List<String> columnList = new ArrayList<>(columns);
            return Arbitraries.of(columnList)
                    .list()
                    .ofMinSize(1)
                    .ofMaxSize(Math.min(5, columnList.size()))
                    .map(fieldNames -> new FieldNameScenario(fieldNames, columns));
        });
    }

    @Provide
    Arbitrary<FieldNameScenario> invalidFieldNameScenarios() {
        return dataTableColumnSets().flatMap(columns ->
                invalidFieldNames(columns)
                        .list()
                        .ofMinSize(1)
                        .ofMaxSize(5)
                        .map(fieldNames -> new FieldNameScenario(fieldNames, columns))
        );
    }

    @Provide
    Arbitrary<FieldNameScenario> mixedFieldNameScenarios() {
        return dataTableColumnSets().flatMap(columns -> {
            List<String> columnList = new ArrayList<>(columns);
            Arbitrary<String> validNames = columnList.isEmpty()
                    ? Arbitraries.strings().alpha().ofMinLength(1).ofMaxLength(20)
                    : Arbitraries.of(columnList);
            Arbitrary<String> invalidNames = invalidFieldNames(columns);

            return Arbitraries.oneOf(validNames, invalidNames)
                    .list()
                    .ofMinSize(1)
                    .ofMaxSize(8)
                    .map(fieldNames -> new FieldNameScenario(fieldNames, columns));
        });
    }

    private Arbitrary<Set<String>> dataTableColumnSets() {
        return columnNames()
                .set()
                .ofMinSize(1)
                .ofMaxSize(10);
    }

    private Arbitrary<String> columnNames() {
        return Arbitraries.strings()
                .alpha()
                .ofMinLength(2)
                .ofMaxLength(20)
                .map(String::toLowerCase);
    }

    private Arbitrary<String> invalidFieldNames(Set<String> validColumns) {
        return Arbitraries.strings()
                .alpha()
                .ofMinLength(1)
                .ofMaxLength(20)
                .filter(s -> !validColumns.contains(s));
    }
}
